### ISSUE TITLE
Custom User Data

### ISSUE THREAD

**Issue by didimitrie** [speckleworks/SpeckleDynamo#32]
Step 0: 
- [x] I've  read the [contribution guidelines](https://github.com/speckleworks/SpeckleOrg/blob/master/CONTRIBUTING.md)!

A lot of nice (*read: core*) functionality in speckle is enabled (in gh for now, and if @PaulPoinet would have finished his plugin, in rhino too /snide) by adding [custom user data](https://speckle.works/doc/userdatautils/) to objects. Specifically, this allows users (clients and humans alike) to ask questions to the API on the likes of `can you please give me all the objects of type polyline with an area in the range of 4 to 10` or `all objects with properties.likes>100`. 

In the world of rhinocommon we're struggling to handle this natively, via `ArchivableDictionaries` which you can set on any `GeometryBase` object (which makes it a pain in grasshopper).

The question is whether it's feasible for dynamo too. I've deleted the issue template, but one section is actually still valid:  

### Proposed solution:

- If possible, implement this as close to native as possible via an equivalent `UserDictionary`. 
- If not, this could be added as a last step in the process before sending data, directly on a SpeckleObject. This entails two extra separate components, ToSpeckle and FromSpeckle, and then extra legwork to create dictionaries that then can be plopped in a SpeckleObject's `properties` field. 

### Extra Context
There is a plan to automatically populate the `properties` field of a SpeckleObject, upon conversion, with simple common sense stuff, like length, area, volume and later on things like centroid, bounding box, etc. or platform specific stuff like layer name, colour, etc.  

There is a further plan to make custom object definitions into something like templates, ie light schema definitions that can be centralised on the speckle server and thereafter magically embedded as object creation tools on the various platforms (ie, automatically create a component that allows users to populate the properties of a line or polyline so that it can be transformed into a wall in archicad or something).

**Comment by teocomi** [comment]
AFAIK DesignScript elements don't have an equivalent UserDictionary, so the second proposed solution might be the way to go. 

@mjkkirschner do you know if a Dictionary field on DesignScript elements for embedding custom meta data might ever be a thing?

**Comment by alvpickmans** [comment]
Seems like there is no proper way to do this at the moment according to this [post on the forum](https://forum.dynamobim.com/t/extensions-how-to-store-custom-parameters-on-ootb-nodes/22821/2).

A work around could be following similar structure as the [Display class](https://github.com/pboyer/Dynamo/blob/master/src/Libraries/CoreNodes/GeometryColor.cs)?
Something like:
```
    public class MetaObject
    {
        internal object item;
        internal Dictionary<string, object> meta;
    }
```

**Comment by mjkkirschner** [comment]
Dynamo geometry types do have a dictionary called “tags” - but not all objects in the vm have this property, only geo.

**Comment by didimitrie** [comment]
>Dynamo geometry types do have a dictionary called “tags” - but not all objects in the vm have this property, only geo.

*noob warning* Can these tags support nested structures? I suspect not, going by the name. Nevertheless, this does sound like a good interim solution to me...

**Comment by alvpickmans** [comment]
> but not all objects in the vm have this property, only geo.

Will there be a case where no geometric objects require custom data?

> noob warning Can these tags support nested structures?

It seems that tags are dictionaries of type `Dictionary<string, object>`, so nested dictionaries should work (?)

**Comment by didimitrie** [comment]
> Will there be a case where no geometric objects require custom data?

not really i guess. ideally yes - any speckle object can accept extra props, but this is not implemented in gh (we would need approach 2 to be able to do it)

>It seems that tags are dictionaries of type Dictionary<string, object>, so nested dictionaries should work (?)

Then it's cool, we can have feature parity with gh!

**Comment by alvpickmans** [comment]
will have a play with it later on this week :+1:

**Comment by radumg** [comment]
> Will there be a case where no geometric objects require custom data?

Maybe not in Dynamo, but there will be in Revit, where plenty of elements have no geometry but would still be highly useful to transfer, from `Project Information` object to weird ones like `Levels`, etc.

**Comment by mjkkirschner** [comment]
I don't think one exists, I would probably use a pre generated key for all speckle - and nest your own dictionary under that.

**Comment by alvpickmans** [comment]
Of course. My only concern about this would be data loss if user adds tags via other node/package rather than speckle. I think Spring package also has nodes to set/read tags. If this is the case, when serialized in Speckle any possible tag not created with SpeckleDynamo won't be sent.

**Comment by didimitrie** [comment]
Would it then not be easier to use the Spring ReadTag and SetTag node then, and just implement the conversion handling in the DynamoConverter then?
